### ISSUE TITLE
[discovery] NOD | Notification capabilities, deadline and decision data

### ISSUE THREAD

**Issue by saderagsdale** [department-of-veterans-affairs/va.gov-team#62597]
## Value Statement
**_As a_** developer
**_I want to_** know what our options are for 1) letting Veterans know the exact decision date of their decision letters, 2) sending notifications to remind Veterans of AMA deadlines, and 3) sending Veterans their decision letters via email.
**_So that_** we can make it easier for Veterans to submit a NOD in a timely and efficient manner.

---
## Background Context

The team just finished [high-level brainstorming and prioritization for NOD](https://app.mural.co/t/coforma8350/m/coforma8350/1684425801227/8a7400acc76c951cc390cceab95bd51edc17bc85?sender=u0d70ac6e87daa7c066838517), and would like to explore the following use cases:

1. A Veteran wants to know the exact decision date when adding a new issue.
2. A Veteran wants to know when their filing deadline is approaching so they can take action.
3. A Veteran wants to have to have their decision letter delivered via email, because decision letters don't always arrive in the mail.

## Out of scope
- Actual code work

## Open questions
- What are our development options for these use cases?

## Tasks

**Use Case 1: A Veteran wants to know the exact decision date when adding a new issue.**
- [ ] Figure out how to supply Veterans with the exact decision date of their decisions.

**Use Case 2: Veteran wants to know when their filing deadline is approaching so they can take action.**
- [ ] Figure out how we might send email and SMS notifications to Veterans.
- [ ] Figure out how we might schedule such notifications to be run at a future date.
- [ ] Figure out whether we are considering sending Veterans multiple reminder notifications, per appeal, and if so, how we might schedule multiple reminder notifications to be run at future dates.
- [ ] Figure out how to supply our scheduled notifications with AMA deadlines.

**Use Case 3: A Veteran wants to have to have their decision letter delivered via email, because decision letters don't always arrive in the mail.**
- [ ] Figure out whether we can send decision letters to Veterans via email

**Other Tasks:**
- [ ] Huddle with team to refine options
- [ ] Talk to Matt and Steve about approach

## Definition of Ready 

- [ ] We're aligned on what we're investigating

## Definition of Done

- [ ] We have a reasonable understanding of what our options are and how each solution would be implemented
- [ ] Options are written into the comments of this ticket
- [ ] We have buy-in from a stakeholder perspective
- [ ] Reviewed and approved by product and/or design

**Comment by rubensun** [comment]
Current list of notifications cataloged by the VA Notify team
Also added as an attachment, so folks can more easily access
[https://app.zenhub.com/files/133843125/f801403b-4f25-468e-82f3-3544e93f498c/download](https://app.zenhub.com/files/133843125/f801403b-4f25-468e-82f3-3544e93f498c/download)

https://docs.google.com/spreadsheets/d/1E7JupFC2i6XUzlOxbWwTbIJZ6AxRUz-m/edit#gid=1231731113

**Comment by rubensun** [comment]
[List of Form-related Notifications](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/form%20confirmations)

**Comment by data-doge** [comment]
Here are my notes so far, per use case, and per task. @kylesoskin feel free to edit this comment (I think you should be able to edit it?)

# Use Case 1: A Veteran wants to know the exact decision date when adding a new issue.

**Context:** In the NOD form, there is an "add issue" page (pictured below), where a Veteran can manually add an issue that was not present in the list of Lighthouse-provided contestable issues on the previous page. From what I understand, the Veteran's issue could be absent from the previous page because: 1) there was an error with Lighthouse contestable_issues endpoint that we use to fetch issues or 2) Lighthouse doesn't have the issue in their system yet. It may also be possible that the contestable_issues endpoint we're using is only returning disability-related issues.

![image](https://github.com/department-of-veterans-affairs/va.gov-team/assets/7520103/724c999f-cc46-4b56-b744-f67140219198)

## Task 1: Figure out how to supply Veterans with the exact decision date of their decisions.

If a Veteran is adding an issue manually, they need to know the exact decision date of the decision on their issue. The first question is whether we could expect the contestable_issues, when working, to return a full list of disability and non-disability issues. If so, then we might consider modifying the UX around contestable_issues endpoint failures (see related: https://github.com/department-of-veterans-affairs/va.gov-team/issues/62872), to encourage the veteran to come back and try again later, rather than manually add their issue. Just a thought. But if it turned out that the contestable_issues endpoint, when working, could only return disability-related issues, then we would have to look elsewhere for non-disability-related issues. We might be able to get that information from the claim status tool, in which case we could link Veterans to it. But if it turned out that the claim status tool had all the data we needed, then we would want to rely on it, instead of the contestable_issues endpoint, alone, on the issues page in NOD. It might make sense to always include a link to the claim status tool on the "add issue" page to handle cases where the contestable_issues endpoint isn't working.

### Next Steps

- [ ] Figure out whether the `contestable_issues` endpoint returns all issues, or just disability-related issues.
- [x] If the latter, figure out whether the claim status tool and its endpoints returns all issues (or at the very least, more than just disability-related issues).
  - Claim status tools' endpoints return all issues, but not their decision dates. Jerek says we _might_ be able to get that from BGS, via the bgs_ext gem.
- [ ] If the claim status tool and its endpoints don't return anything more than the `contestable_issues` endpoint, briefly investigate whether alternative data sources exist.

# Use Case 2: Veteran wants to know when their filing deadline is approaching so they can take action.

**Context:** After a decision is made on a claim, Veterans have 90 days to file an NOD. We want to send email and SMS reminders about this deadline.**

## Task 2: Figure out how we might send email and SMS notifications to Veterans.

VANotify appears to be _the_ way to send email, SMS, (and push), notifications to Veterans. They have a lot of great documentation. Good reading:
- [VANotify Playbook](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/va-notify/VA%20Notify%20Playbook.pdf) (Tells you how to set up VANotify for your application)
- [VANotify Product README](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/va-notify/README.md)
- [VANotify Notification Guide](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/va-notify/notification-guide.md)
- [VANotify 2023 Roadmap](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/va-notify/Goals%20and%20roadmap%202023.md)

To use VANotify, we would follow the steps in the playbook linked above. Notably, VANotify's 2023 roadmap contemplates "add[ing] support for Twilio notification status callbacks," that I presume would allow us to handle failed SMS deliveries. An open question I have is whether we would need VANotify to obtain a new short-code phone number for us. I'm guessing we could just use the same short-code phone number that I presume is shared by all of VA.gov. If we couldn't, then it could take up to 3 months to obtain a shortcode phone number.

### Next Steps

- [ ] Check whether VANotify Twilio notification status callbacks have been set up yet.
- [ ] Check whether we would need to obtain a new short-code phone number if were to use VANotify to set up NOD reminders.

## Task 3: Figure out how we might schedule such notifications to be run at a future date.

Reminders need to be scheduled, so this is something we need to figure out. This is might be pretty tricky. Here are some options:

**Option 1: VA Event Bus**

There is a team that is currently working on a VA Event Bus that would receive events from VBMS (including, potentially, "Decision Letter Created" events) and publish them to subscribers. Ideally, we could subscribe to any events relating to the creation/filing/delivery of a decision letter, and upon receiving such an event and when appropriate, enqueue a Sidekiq job to send an email and SMS message to the Veteran. This solution would depend on a number of factors, including:
- Whether the VA Event Bus is ready to use
- Whether it would receive the events we need from VBMS
- Whether we would be able to subscribe and filter down to only the events relevant to our use case.
- Whether the event would contain all the information we need to send our notifications (e.g. decision date; issue name; Veteran's phone and email, or the means to programmatically obtain the Veteran's phone and email)

**Option 2: ???**

### Next Steps

- [ ] Check on the status of VA Event Bus.
- [ ] Figure out whether VA Event Bus, whenever it is available, could give us the information we need for our use case.
- [ ] Think about alternatives.

## Task 4: Figure out whether we are considering sending Veterans multiple reminder notifications, per appeal, and if so, how we might schedule multiple reminder notifications to be run at future dates.

TODO

### Next Steps

TODO

## Task 5: Figure out how to supply our scheduled notifications with AMA deadlines.

TODO

### Next Steps

TODO

# Use Case 3: A Veteran wants to have to have their decision letter delivered via email, because decision letters don't always arrive in the mail.

## Task 6: Figure out whether we can send decision letters to Veterans via email

(Rocio says she already gets decision letters via email)

**Comment by saderagsdale** [comment]
@data-doge review and let us know when you're ready to chat.

**Comment by data-doge** [comment]
@saderagsdale I haven't had a chance to review this ticket yet, but it's on my todo list

**Comment by data-doge** [comment]
Quick notes for tomorrow's conversation with Eileen and Sade:

**Use Case 1:**
- We can get an approximate decision date (`approxDecisionDate`) from the contestable issues returned by Lighthouse's `contestable_issues` endpoint. But we need to figure out how approximate this date is, and whether it would warrant a disclaimer.
- Currently, we are calling Lighthouse's `contestable_issues` endpoint with a `benefit_type` of `compensation`. We need to figure out whether this endpoint returns the contestable issues that we expect the NOD form to encompass. That means we need to figure out exactly what benefits the NOD form should and should not encompass, and what Lighthouse `benefit_type`s they map to. Depending on what we learn, we may need to call `contestable_issues` multiple times, with different `benefit_type`s.

**Use Case 2:**
- The notification work is going to be very complicated and time-consuming. We most likely need an event bus. There is a VA team working on building an event bus, but they are too early in their development for us to consider them as an option. We could try building our own custom solution, but we likely won't have the time and resources for that.
- We can, however, set up notifications to remind veterans to complete in progress forms. VANotify has an infrastructure for that, that we can opt in to.

**Use Case 3:**

- Rocio says that she already receives decision letters via email.

**Comment by data-doge** [comment]
Draft notes for me to refine later:

- Appx decision date, how approximate. Ask LH. Ask Emily. Important for rendering, but also for notification work. Would be helpful to get real example responses. Possibly from vets on our team.
- Scope of NOD and LH endpoints. Compensation endpoint, believe encompasses 80%. Unsure whether encompasses disability or non-disability claims only. We could also just make requests for all ten benefit_types provided by LH. Good argument for that: ENG work might just be a 5 pt ticket; help resolve duplicate claims problem; avoids content/disclaimer problem; also potential to bring this expansion into the other DR forms. Need to think more about unintended consequences that could result from such an expansion. One thought, from an engineering perspective, if we make 10 async requests on the client, we need to think about how to deal with failed requests and receiving incomplete lists.
- Re: scheduled notifications. Event bus team wrapping up MVP sometime in 2024. Uncertainty there, but at the very least, MVP includes decision dates. May be able to use existing "VIP" event bus directly. Kyle can look into this. VBMS responsible for updating decision dates. Possibly other systems as well. We can reach out to these systems to see if they would be open to consuming a webhook we provide, or providing a query endpoint for us to consume. For VBMS, can talk to Emily to get eng contact there.
- No matter what, can still set up inprogress notifications with VA notify. If we do, may want to include email opt-in copy in save in progress page.
- As for DLs by email, need to figure out whether self-select, or default. If the latter, should add opt-in content to confirmation page. Will also work with vets on our team to understand how to opt-in to notifications.
- For sprint hygiene, will consider whether to inflate estimate on this ticket, or create separate spikes for FU investigations.

**Comment by saderagsdale** [comment]
@data-doge @kylesoskin I'm weeding out complexity and getting back to the purpose of the ticket. Tasks for each use case and action items below:

**A Veteran wants to know the exact decision date when adding a new issue.**
- Action item 1: Find the easiest way to determine whether this is available through BGS, via the bgs_ext gem.
- Action item 2: Confirm with Michael Hobson is the other benefit types are ready to use or need to be reviewed to confirm if they meet business requirements first. 
- Out of scope: over-indexing on adding benefit types. If it's there, great, if not, park it. 

**A Veteran wants to know when their filing deadline is approaching so they can take action.**
- Action item 3: Find out where filing deadline data is stored. Possibly BGS?
- Action item 4: Find out if BIP is useable as an email event bus (or another mechanism). If not. What other ways could we put deadline info in front of Veterans?

**A Veteran wants to have to have their decision letter delivered via email, because decision letters don't always arrive in the mail.**
- Action item 5: Find out if we can parse or otherwise surface and data from decision letters in the form flow.

I am going to look for leads on action items 1-4. @data-doge can you look into action item 5?

**Comment by saderagsdale** [comment]
Sent a note to Premal on action items 1-4.

**Comment by saderagsdale** [comment]
Found BIP API and Swagger documentation here, but you need to be in CAG to click the links:
https://github.ec.va.gov/EPMO/bip-vetservices-claims/blob/development/bip-vetservices-claims-docs/new-integration-partners.md
https://claims-test.dev8.bip.va.gov/swagger-ui.html#/

**Comment by saderagsdale** [comment]
Also got a link to this [developer documentation](https://github.com/department-of-veterans-affairs/abd-vro/wiki) for the VRO platform from Premal, which may have some of the data and capabilities we're looking for.

**Comment by saderagsdale** [comment]
Scheduled time to discuss action items 1-4 with Diana from VRO this Friday.

**Comment by HeatherWidmont** [comment]
@data-doge is planning on breaking this ticket down into 3 separate tickets

**Comment by saderagsdale** [comment]
@data-doge Can you tag me in this ticket when you've looked at the documentation Diana from VRO shared? I'm looking to understand how we can quickly confirm if VRO has what we need.

**Comment by data-doge** [comment]
Hey @saderagsdale sure thing. I plan on doing that today.

Also, after our conversation, and after speaking with Eileen and Ruben today, I've broken this ticket into three separate, focused spike tickets that can be worked on in parallel. The VRO-related ticket is the number one priority:
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/66231 (assigned to Robin)
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/66238 (assigned to me - this is the VRO-related ticket)
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/66251 (assigned to Robin)

Accordingly, I'll remove this ticket from the sprint doc, and slot in these three smaller tickets instead.
